## Q.1
The formula for Euclidean distance is $d(p,q) = \sqrt{\,\sum_{i=1}^n (q_i - p_i)^2}$

With these x values below we find the distances between x1 and x2 to x3. 

In [21]:
x1 = (1,-1)
x2 = (-1,1)
x3 = (-2,3)

In [22]:
dist1 = ((-2-1)**2 + (3--1)**2)**0.5
print("distance 1 = " + str(dist1))

dist2 = ((-2--1)**2 + (3-1)**2)**0.5
print("distance 2 = " + str(dist2))

distance 1 = 5.0
distance 2 = 2.23606797749979


It would classify x3 as class y2 as the euclidean distance is closest to point x2 at 2.24 compared to x1 at 5.0.

## Q.2
When K is odd in a KNN algorithm we do not need a tie-breaking policy because one class will always have a greater number of points closer to the new point than the other class. For example if K=5, the closest you can get to a tie will be 3 points of class y1 and 2 points of class y2 being the nearest neighbours, in this situation and with all odd K numbers the point will always be designated as the class with the highest number of neighbours, therefore it would be classed as label y1. 

## Q.3
A classifier that has an accuracy of 99.9% can be terrible for some datasets if 0.1% of the data is another class. This algorithm will label every data point as the same class and therefore miss the 0.1% that is different. While this appears to give a very high accuracy, it is actually not truly classifying anything rather than just assigning the same class to all data points. This can be extremely bad if misidentifying that 0.1% of data results in a far greater cost than identifying 99.9% of the other data.  

## Q.4
A precision of 1.0 and low recall of 0.1 means that out of the data you predicted positive, all were correctly identified as true positive. However, the low recall means that you actually missed a lot of the true positives and mislabelled them as false negatives. Therefore, when it detects a positive it can be trusted but it is far less trustworthy when detecting negatives as there is a high chance they could be positive. In terms of this classifier, it should not be trusted if it detects that point to not belong to class y and it should be trusted if it detects it to belong to class y. 

## Q.5
The K=1 classifier struggled the most with classes 4 and 9, correctly identifying 40 4s but misidentifying 4 of them (10%) as 9s. 

## Q.6

In [23]:
import gzip
import pickle

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('darkgrid')

# Selecting the training data from the original dataset
f = gzip.open('mnist.pkl.gz', 'rb')
X, y = pickle.load(f, encoding='latin1')[0]
f.close()

# Subsampling
sample_size = 2000
X, y = X[:sample_size], y[:sample_size]

In [24]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [25]:
from sklearn import svm
clf = svm.SVC()
clf.fit(X_train,y_train)
print('Test dataset accuracy: ' + str(clf.score(X_test,y_test)))

Test dataset accuracy: 0.9275


## Q.7

In [48]:
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier

parameters = {'n_estimators': [50,100,200], 'max_features': [0.1,0.25]}

rfc = RandomForestClassifier()
rfc_cv = GridSearchCV(rfc, parameters, cv=5)
rfc_cv.fit(X_train, y_train)
print('Best hyperparameter settings: {0}.'.format(rfc_cv.best_params_))
print('Average accuracy across folds of best hyperparameter setting: {0}.'.format(rfc_cv.best_score_))
print('Test dataset accuracy of best hyperparameter setting: {0}.'.format(rfc_cv.score(X_test, y_test)))

Best hyperparameter settings: {'max_features': 0.1, 'n_estimators': 200}.
Average accuracy across folds of best hyperparameter setting: 0.9106250000000001.
Test dataset accuracy of best hyperparameter setting: 0.9075.
